Add VQGAN-CLIP research project

ErwannMillon opened the pull request:
# What does this PR do?
Implements VQGAN-CLIP using huggingface CLIP models

Related to #21064 

This Research Project allows users to generate or edit images with a single line of code. It wraps the huggingface CLIPProcessor class, allowing images to be processed as torch tensors in order to preserve gradient flow through the transformations. 

Features:
- Positive and negative prompts
- Multiple prompts
- Prompt Weights
- Creating GIF animations of the transformations
- Wandb logging

Tagging @amyeroberts for review :)

Commit:
Add VQGAN-CLIP research project

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_21329). All of your documentation changes will be reflected on that endpoint.

Commit:
fixed style issues

amyeroberts commented on examples/research_projects/vqgan-clip/loaders.py:
```diff
@@ -0,0 +1,75 @@
+import importlib
+
+import torch
+
```
nit 
```suggestion
```

amyeroberts commented on examples/research_projects/vqgan-clip/img_processing.py:
```diff
@@ -0,0 +1,51 @@
+    return x
+
+
+def get_pil(x):
```
nit: function name is a bit misleading as it doesn't return a PIL image. The logic is also repeated in `custom_to_pil`.

amyeroberts commented on examples/research_projects/vqgan-clip/VQGAN_CLIP.py:
```diff
@@ -0,0 +1,266 @@
+        if self.log:
+            self._init_logging(pos_prompts, neg_prompts, image_path)
+
+        assert pos_prompts, "You must provide at least one positive prompt."
```
This should be an exception as it's checking inputs - asserts are there for checking bugs in the code 🐛

amyeroberts commented on examples/research_projects/vqgan-clip/VQGAN_CLIP.py:
```diff
@@ -0,0 +1,266 @@
+            similarity_logits = similarity_logits * weights
+        return similarity_logits.sum()
+
+    def _get_CLIP_loss(self, pos_prompts, neg_prompts, image):
```
To be consistent with naming above

```suggestion
    def _get_clip_loss(self, pos_prompts, neg_prompts, image):
```

amyeroberts reviewed (commented):
Nice - looking good! Thanks for adding this example. 

I've made a few comments, mostly nits. None are critical and I'll leave it to you to decide what you do or don't want to change. 

One thing to update is the `face.jpeg` image. To avoid adding lots of data to the repo, could you do the same as in other examples, and upload the image to a dataset where it can be display through hrefs and downloaded from e.g. https://github.com/huggingface/transformers/blob/42b60f8b02941b0c40c42e150a101eb372c3856e/examples/research_projects/codeparrot/README.md?plain=1#L3 ?

ErwannMillon committed:
Update examples/research_projects/vqgan-clip/README.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ErwannMillon committed:
Update examples/research_projects/vqgan-clip/VQGAN_CLIP.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ErwannMillon committed:
Update examples/research_projects/vqgan-clip/requirements.txt

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ErwannMillon committed:
Update examples/research_projects/vqgan-clip/README.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ErwannMillon committed:
Update examples/research_projects/vqgan-clip/VQGAN_CLIP.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ErwannMillon committed:
Update examples/research_projects/vqgan-clip/VQGAN_CLIP.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ErwannMillon committed:
Update examples/research_projects/vqgan-clip/VQGAN_CLIP.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ErwannMillon committed:
Update examples/research_projects/vqgan-clip/loaders.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ErwannMillon committed:
replace CLIPProcessor with tokenizer, change asserts to exceptions

ErwannMillon committed:
Merge branch 'vqganclip' of github.com:ErwannMillon/transformers into vqganclip

ErwannMillon committed:
rm unused import

ErwannMillon commented:
Hi there,
Thanks for the feedback, I fixed the style issues and removed the face image. 
Also refactored the code to have more accurate function names, used just the tokenizer, changed the assertions to exceptions, and removed some extraneous code (eg double crop, freeze_module)
Have a great day :)
Erwann

amyeroberts commented:
@ErwannMillon - thanks for the updates. It's looking good 😎 ! 

Just two last things before I think we're ready to merge: 
* Removing the `face.jpg` file and instead pointing to a place where it can be downloaded
* Could you also remove the notebook? Like the image, we want to avoid adding large files as much as possible. Happy for you to link to e.g. a colab which shows this demo in the README.

ErwannMillon committed:
remove large files (jupyter notebook linked in readme, imgs migrated to hf dataset)

amyeroberts commented on examples/research_projects/vqgan-clip/VQGAN_CLIP.py:
```diff
@@ -0,0 +1,269 @@
+
+    def __init__(self, device: str = "cpu", clip_model: str = "openai/clip-vit-large-patch14") -> None:
+        self.device = device
+        self.tokenizer = CLIPTokenizerFast.from_pretrained(clip_model)
```
If using the fast tokenizers, you'll need to add the `tokenizers` library as a requirement

ErwannMillon replied:
Done

ErwannMillon committed:
add tokenizers dependency

amyeroberts reviewed (approved):
Looks good to merge from my end :)

ErwannMillon committed:
Remove comment

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

sgugger reviewed (approved):
Thanks for working on this! Can we just remove the empty folder `model_checkpoints` from the PR?

ErwannMillon committed:
rm model checkpoints
